stronghold/stronghold.gd

```diff
@@ -1,44 +1,82 @@
 class_name Stronghold
 extends Node2D
 
+@onready var occupants: Node = $Occupants
+@onready var sprites: Node2D = $Sprites
+
 var _health: Health
+var _level: int = 0
+var _max_level: int = 2
+var _base_health: int = 100
+var _current_occupants: int = 0
+var _occupant_slot_progression: Array[int] = [5, 12, 21]
 
 
 func _ready() -> void:
-	GunmanUpgrade.new().level_changed.connect(_handle_gunman_upgrade)
-	_health = Health.new(100)
+	_health = Health.new(_base_health)
 	_health.health_changed.connect(SignalBus.player_health_changed.emit)
 	_health.died.connect(SignalBus.player_died.emit)
 
+	GunmanUpgrade.new().level_changed.connect(_handle_gunman_upgrade)
+	StrongholdUpgrade.new().level_changed.connect(_handle_stronghold_upgrade)
 	RepairmanUpgrade.new().level_changed.connect(_on_repairman_upgraded)
 	RepairUpgrade.new(_health).level_changed.connect(_on_repair_level_changed)
 	SignalBus.player_hit.connect(_on_player_hit)
 
+	self._update_sprite()
+
 
 func get_health() -> Health:
 	return _health
 
 
-func _handle_gunman_upgrade(_level: int):
-	var gunman = Gunman.create()
-	var posts = get_tree().get_nodes_in_group("gunman_posts") as Array[Node2D]
-	var min_occupancy = (
-		posts
-		. map(func(p): return p.get_child_count())
-		. reduce(func(min, value): return value if value < min else min)
-	)
-
-	(
-		posts
-		. filter(func(p): return p.get_child_count() <= min_occupancy)
-		. pick_random()
-		. add_child(gunman)
-	)
+func _handle_stronghold_upgrade(_level: int):
+	self._level = _level
+	self._update_sprite()
+	self._check_vacancy()
+	if _level >= self._max_level:
+		SignalBus.stronghold_max_level_reached.emit()
 
+
+func _max_occupant_slots() -> int:
+	return self._occupant_slot_progression[self._level]
+
+
+func _handle_gunman_upgrade(_level: int):
 	var target_areas = get_tree().get_nodes_in_group("enemy_spawn_areas") as Array[CollisionShape2D]
 	var area: CollisionShape2D = target_areas.pick_random()
+
+	var gunman = Gunman.create()
 	gunman.rotation = gunman.global_position.angle_to_point(area.global_position)
```

Comment:
Maybe doesn't need to be in this PR, but would this "point to enemy" behavior not be better suited to have in the _ready of the gunman itself?

Reply:
yeah it would